File: 2115-find-all-possible-recipes-from-given-supplies/2115-find-all-possible-recipes-from-given-supplies.java
import java.util.*;

class Solution {
    public List<String> findAllRecipes(String[] recipes, List<List<String>> ingredients, String[] supplies) {
        Set<String> supplySet = new HashSet<>(Arrays.asList(supplies));
        Map<String, List<String>> graph = new HashMap<>();
        Map<String, Integer> inDegree = new HashMap<>();

        for (int i = 0; i < recipes.length; i++) {
            for (String ing : ingredients.get(i)) {
                if (!supplySet.contains(ing)) {
                    graph.computeIfAbsent(ing, k -> new ArrayList<>()).add(recipes[i]);
                    inDegree.merge(recipes[i], 1, Integer::sum);
                }
            }
        }

        Queue<String> q = new ArrayDeque<>();
        for (String r : recipes) if (inDegree.getOrDefault(r, 0) == 0) q.offer(r);
        
        List<String> result = new ArrayList<>();
        while (!q.isEmpty()) {
            String curr = q.poll();
            result.add(curr);
            for (String nxt : graph.getOrDefault(curr, Collections.emptyList())) {
                if (inDegree.merge(nxt, -1, Integer::sum) == 0) q.offer(nxt);
            }
        }
        return result;
    }
}

File: 2115-find-all-possible-recipes-from-given-supplies/2115-find-all-possible-recipes-from-given-supplies.py
class Solution:
    def findAllRecipes(self, recipes: List[str], ingredients: List[List[str]], supplies: List[str]) -> List[str]:
        available = set(supplies)
        ans, ingredient_to_recipe, in_degree = [], defaultdict(set), Counter()
        for rcp, ingredient in zip(recipes, ingredients):
            non_available = 0
            for ing in ingredient:
                if ing not in available:
                    non_available += 1
                    ingredient_to_recipe[ing].add(rcp)
            if non_available == 0:
                ans.append(rcp)
            else:
                in_degree[rcp] = non_available
        for rcp in ans:
            for recipe in ingredient_to_recipe.pop(rcp, set()):
                in_degree[recipe] -= 1
                if in_degree[recipe] == 0:
                    ans.append(recipe)
        return ans    